File: src/Search/Binary.php
<?php

namespace GulDmitry\Algorithms\Search;

/**
 * O(log n)
 */
class Binary
{
    /**
     * @param array $searchIn
     * @param int $num
     * @return int|null Position.
     */
    public function search(array $searchIn, int $num): ?int
    {
        sort($searchIn);
        $lowInd = 0;
        $highInd = count($searchIn) - 1;

        while ($lowInd <= $highInd) {
            $midInd = (int)round(($lowInd + $highInd) / 2);
            $midElm = $searchIn[$midInd];

            if ($midElm === $num) {
                return $midInd;
            } elseif ($midElm < $num) {
                $lowInd = $midInd + 1;
            } elseif ($midElm > $num) {
                $highInd = $midInd - 1;
            }
        }
        return null;
    }
}

File: src/go/Search/binary.go
package Search

import "sort"

// log(N)
// -1 if element is not found.
func Binary(input []int, num int) (position int) {
	position = -1
	sort.Ints(input)
	lowInd := 0
	highInd := len(input) - 1

	for lowInd <= highInd {
		midInd := (lowInd + highInd) / 2
		midElm := input[midInd]

		if midElm == num {
			position = midInd
			return
		} else if midElm > num {
			highInd = midInd - 1
		} else if midElm < num {
			lowInd = midInd + 1
		}
	}
	return
}
